[ES|QL] Allow DateTime as the third and fourth inputs to auto_bucket

fang-xing-esql opened the pull request:
Resolves: #102756 #104646 

Currently, the AUTO_BUCKET function takes only String data type for the start and end time range. This enhances the AUTO_BUCKET function by allowing data with DateTime type as input to the start and end time range, users do not need to convert the input parameters from DateTime to String type manually.

The AUTO_BUCKET function does not allow variables as input to the start and end time range yet, this will be addressed in a separate PR.

fang-xing-esql committed:
Allow both string and datetime as the third and fourth inputs to auto_bucket

 Committer: Fang Xing <fang.xing@elastic.co>

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucketTests.java:
```diff
@@ -80,12 +80,13 @@ public static Iterable<Object[]> parameters() {
         })));
     }
 
+    @SuppressWarnings("checkstyle:LineLength")
```
We should just run the formatter I think.

fang-xing-esql replied:
done

nik9000 replied:
Could you remove this suppression as well? That'll make sure it never grows too long accidentally.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java:
```diff
@@ -214,6 +220,36 @@ private TypeResolution resolveType(BiFunction<Expression, TypeResolutions.ParamO
+    private long convertToLong(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        long v = 0;
+        TypeResolution resolution = isDate(e, operationName, paramOrd);
+        if (resolution.resolved()) {
```
Since we've already resolved the types on the subfields we should be able to do `if (e.dataType() == DataTypes.DATE)`. We shouldn't need the resolution thingy here.

fang-xing-esql replied:
done

nik9000 commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/TypeResolutions.java:
```diff
@@ -170,7 +170,7 @@ public static TypeResolution isType(
     }
 
-    private static String acceptedTypesForErrorMsg(String... acceptedTypes) {
+    public static String acceptedTypesForErrorMsg(String... acceptedTypes) {
```
I imagine we should move the utility function into the type rather than make this public.

fang-xing-esql replied:
done

nik9000 reviewed (changes requested):
Looks about right! I've left a few comments asking to move some methods around and run the formatter, but it does seem to get the job done!

fang-xing-esql committed:
Merge branch 'elastic:main' into main

fang-xing-esql committed:
Allow both string and datetime as the third and fourth inputs to auto_bucket

fang-xing-esql commented:
> Looks about right! I've left a few comments asking to move some methods around and run the formatter, but it does seem to get the job done!

Thanks for reviewing! All of the comments are addressed.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java:
```diff
@@ -115,8 +115,9 @@ public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEv
-            long t = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(((BytesRef) to.fold()).utf8ToString());
+            long f = convertToLong(from, sourceText(), THIRD);
+            long t = convertToLong(to, sourceText(), FOURTH);
+            ;
```
leftover `;`

fang-xing-esql replied:
This is a good catch!

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucketTests.java:
```diff
@@ -80,12 +80,17 @@ public static Iterable<Object[]> parameters() {
+                Source.EMPTY,
+                DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis("2023-03-01T09:00:00.00Z"),
+                DataTypes.DATETIME
+            );
```
I just realized you can mix and match these! Neat.

I think it'd probably be nice to do something like:
```
private Literal fromOrTo(String date) {
  if (randomBoolean() {
    return new Literal(Source.EMPTY, new BytesRef(date), randomBoolean() ? DataTypes.KEYWORD : DataTypes.TEXT);
  }
  return new Literal(
                Source.EMPTY,
                DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(date),
                DataTypes.DATETIME
            )
}
```

and then call it for both the `from` and `to`. It'll make a random valid endpoint every time. If you stick the `@Repeat(iterations=100)` annotation on the class it'll run the whole thing 100 times. You can't commit that, but it's fine - just run it a few times locally to make sure all the random combinations work. Then remove the annotation and rerun the formatter.

From there on out the shear number of people and CI runs that run this will reproduce all the valid cases fairly frequently.

fang-xing-esql replied:
That's a good idea!

nik9000 commented:
@fang-xing-esql I've made two more small suggestions. That second one about randomization should be fun to learn about. It's quite useful and the way we randomize our tests is fairly unique. Very not-quickcheck. Very low overhead at write and runtime but kind of annoying to reproduce. It works ok though!

fang-xing-esql committed:
Allow both string and datetime as the third and fourth inputs to auto_bucket

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/TypeResolutions.java:
```diff
@@ -118,6 +118,25 @@ public static TypeResolution isIPAndExact(Expression e, String operationName, Pa
         return isExact(e, operationName, paramOrd);
     }
 
+    public static TypeResolution isStringOrDate(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
```
It's rarely a good idea to touch the QL classes.
In this case instead of adding a specialized method in QL (which is used just once), use the following one liner in AutoBucket relies on `TypeResolutions#isType()` method (see the declarations of isDate/isIP/isString/ etc...:
```
TypeResolutions.isType(e, exp -> { DataTypes.isString(exp) || DataTypes.isDateTime(exp) }, operationName, paramord, "datetime", "string');
```

costin replied:
As a side-note, QL projects rely on a centralize pluggable utility to perform conversions yet this is not leverages through-out the ESQL codebase (see `EsqlDataTypeConverter`) - this needs to change but in separate PR for the future https://github.com/elastic/elasticsearch/issues/104596

fang-xing-esql replied:
Moved isStringOrDate from TypeResolutions to AutoBucket.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java:
```diff
@@ -216,6 +216,16 @@ private TypeResolution resolveType(BiFunction<Expression, TypeResolutions.ParamO
         return isFoldable(to, sourceText(), FOURTH);
     }
 
+    private long convertToLong(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        long v = 0;
+        if (e.dataType() == DataTypes.DATETIME) {
+            v = ((Number) e.fold()).longValue();
+        } else {
+            v = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(((BytesRef) e.fold()).utf8ToString());
+        }
+        return v;
+    }
```
Only expression e is needed, the other params are not.
The method itself can also be tighten using the ternary operators and existing utilities:
```
val value = Foldables.valueOf(e);
return DataTypes.isDate(e) ? ((Number) value).longValue() : DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(BytesRefs.toString(value));
```

costin reviewed (changes requested):
Left a couple of comments.

fang-xing-esql committed:
Allow both string and datetime as the third and fourth inputs to auto_bucket

fang-xing-esql committed:
Allow both string and datetime as the third and fourth inputs to auto_bucket

fang-xing-esql commented:
> Left a couple of comments.

Thank you for the suggestions!

costin reviewed (approved):
LGTM

costin reviewed (commented):
Forgot to mention - please add some basic csv tests, next to the existing auto-bucket ones.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java:
```diff
@@ -216,6 +216,24 @@ private TypeResolution resolveType(BiFunction<Expression, TypeResolutions.ParamO
         return isFoldable(to, sourceText(), FOURTH);
     }
 
+    public static TypeResolution isStringOrDate(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
```
Nit, style preference: this could be inlined, potentially by adding an Expression.TypeResolution#or() method, similar to the existing [`and()`](https://github.com/elastic/elasticsearch/blob/a1344106f59a03eb455e73309a537dc2e3133042/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/Expression.java#L53).

But it can also stay as is.

fang-xing-esql replied:
I'll keep it here for now, as only AutoBucket uses it. If there is other caller found in the future, factor it into Expression.TypeResolution will be better.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java:
```diff
@@ -216,6 +216,24 @@ private TypeResolution resolveType(BiFunction<Expression, TypeResolutions.ParamO
+    public static TypeResolution isStringOrDate(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        return TypeResolutions.isType(
+            e,
+            exp -> { return DataTypes.isString(exp) || DataTypes.isDateTime(exp); },
```
```suggestion
            exp -> DataTypes.isString(exp) || DataTypes.isDateTime(exp),
```

fang-xing-esql replied:
Done.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java:
```diff
@@ -216,6 +216,24 @@ private TypeResolution resolveType(BiFunction<Expression, TypeResolutions.ParamO
+        );
+    }
+
+    private long convertToLong(Expression e) {
```
IMO `convertToLong` suggests that the expression is going to be interpreted as long, not folded to long. It'd be good to convey that intention too. Maybe `foldToLong()`, `foldToMillis()` or `foldAsDatetime()` (it "just so happens" that datetime is encoded as a long) or along these lines.

fang-xing-esql replied:
Renamed, thank you for the suggestion!

bpintea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucketTests.java:
```diff
@@ -84,15 +84,22 @@ private Expression build(Source source, Expression arg) {
         return new AutoBucket(source, arg, new Literal(Source.EMPTY, 50, DataTypes.INTEGER), from, to);
     }
 
+    private Literal fromOrTo(String date) {
```
Should this be `stringOrDatetime()`?

fang-xing-esql replied:
Renamed, thank you for the suggestion!

bpintea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucketTests.java:
```diff
@@ -84,15 +84,22 @@ private Expression build(Source source, Expression arg) {
     }
 
+    private Literal fromOrTo(String date) {
+        if (randomBoolean()) {
```
Nit, style preference: opportunity to use also here the ternary operator. But might break the lines making it harder to read.

bpintea reviewed (approved):
LG, only left minor notes.

fang-xing-esql committed:
Allow both string and datetime as the third and fourth inputs to auto_bucket

fang-xing-esql commented:
> Forgot to mention - please add some basic csv tests, next to the existing auto-bucket ones.

Added a new test case, and modified a few existing test cases in CsvTest to allow auto_bucket to take DateTime as input.

fang-xing-esql commented:
> LG, only left minor notes.

Thank you for reviewing, all of the comments are addressed.

fang-xing-esql committed:
Merge branch 'elastic:main' into main

fang-xing-esql commented:
@elasticmachine run buildkite/docs-build-pr

fang-xing-esql commented:
@elasticmachine run buildkite/docs-build-pr

brianseeders commented:
run docs-build
